src/IW/Sync/Update.hs

```diff
@@ -4,37 +4,82 @@ data and insert it into the database.
 -}
 
 module IW.Sync.Update
-       ( fetchAndUpsertRepos
+       ( syncRepos
        ) where
 
 import Control.Monad.IO.Unlift (MonadUnliftIO)
+import Data.Time (Day (..), addDays)
 import UnliftIO.Async (mapConcurrently_)
 
 import IW.App (WithError)
 import IW.Core.Repo (Repo (..))
 import IW.Db (WithDb, upsertRepos, updateRepoCategories)
 import IW.Effects.Cabal (MonadCabal (..), getCabalCategories)
-import IW.Sync.Search (fetchAllHaskellRepos, fromGitHubRepo)
+import IW.Sync.Search (searchHaskellReposByDate, fromGitHubRepo, liftGithubSearchToApp)
 
 
--- | This function fetches the latest repos from the GitHub API, parses their @.cabal@ files,
--- and upserts them into the database.
-fetchAndUpsertRepos
+-- | This function fetches repos from the GitHub API within a specified date range,
+-- parses their @.cabal@ files, and upserts them into the database.
+syncRepos
     :: forall env m.
        ( MonadCabal m
        , MonadUnliftIO m
        , WithDb env m
        , WithLog env m
        , WithError m
        )
-    => m ()
-fetchAndUpsertRepos = do
-    gitHubRepos <- fetchAllHaskellRepos
+    => Day     -- ^ Oldest day
+    -> Day     -- ^ Most recent day
+    -> Integer -- ^ Interval of days before most recent day
+    -> Int     -- ^ Page
+    -> m ()
+syncRepos oldest recent interval page =
+    if recent == oldest
+    then log I $ "Oldest day " <> julianDayToIso oldest <> " reached..."
+    else sync
+  where
+    sync :: m ()
+    sync = do
+        resCount <- syncReposByDate intervalStart recent page
+        if resCount < 100
+            then syncRepos oldest nextRecent interval 1
+            else syncRepos oldest recent interval (page + 1)
+
+    intervalStart :: Day
+    intervalStart = negate interval `addDays` recent
+
+    nextRecent :: Day
+    nextRecent = pred intervalStart
```

Comment:
Yeah, the HLint rule in `relude` should be fixed in that case. For now you can just ignore the rule inside this module.